primitives/numeric/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+        };
+
+        Ok(Self { inner })
+    }
+
+    /// Return mantissa of number
+    /// E.g.
+    /// - 100 (scale 0) mantissa is 100
+    /// - 1.01 (scale 2) mantissa is 101
+    /// - 0.042 (scale 3) mantissa is 3
+    #[inline]
+    pub const fn mantissa(&self) -> u128 {
+        // Non-negative invariant
+        #[allow(clippy::cast_sign_loss)]
+        {
+            self.inner.mantissa() as u128
+        }
+    }
+
+    /// Return scale of number
+    #[inline]
+    pub const fn scale(&self) -> u32 {
+        self.inner.scale()
+    }
+
+    /// Checked addition
+    ///
+    /// # Errors
+    /// In case of overflow
+    pub fn checked_add(self, other: Self) -> Option<Self> {
+        self.inner
+            .checked_add(other.inner)
+            .map(|inner| Self { inner })
+    }
+
+    /// Checked subtraction
+    ///
+    /// # Errors
+    /// In case of overflow
+    pub fn checked_sub(self, other: Self) -> Option<Self> {
+        self.inner
+            .checked_sub(other.inner)
+            .and_then(|inner| inner.is_sign_positive().then_some(Self { inner }))
+    }
+
+    /// Check if number is zero
+    pub const fn is_zero(&self) -> bool {
+        self.inner.is_zero()
+    }
+}
+
+/// Convert [`Numeric`] to `f64` with possible loss in precision
+impl From<Numeric> for f64 {
+    fn from(value: Numeric) -> Self {
+        value.inner.to_f64().expect("never fails")
+    }
+}
+
+impl NumericSpec {
+    /// Check if given numeric satisfy constrains
+    ///
+    /// # Errors
+    /// If given number has precision or scale higher than specified by spec.
+    pub fn check(self, numeric: &Numeric) -> Result<(), NumericSpecError> {
+        if !self.scale.map_or(true, |scale| scale >= numeric.scale()) {
+            return Err(NumericSpecError::ScaleTooHigh);
+        }
+
+        Ok(())
+    }
+
+    /// Create [`NumericSpec`] which accepts any numeric value
+    #[inline]
+    pub const fn unconstrained() -> Self {
+        NumericSpec { scale: None }
+    }
+
+    /// Create [`NumericSpec`] which accepts only integer values
+    #[inline]
+    pub const fn integer() -> Self {
+        Self { scale: Some(0) }
+    }
+
+    /// Create [`NumericSpec`] which accepts numeric values with scale up to given decimal places
+    #[inline]
+    pub const fn fractional(scale: u32) -> Self {
+        Self { scale: Some(scale) }
+    }
+}
+
+impl core::str::FromStr for Numeric {
+    type Err = NumericError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Ok(inner) = Decimal::from_str(s) else {
+            return Err(NumericError::Malformed);
+        };
+
+        if inner.is_sign_negative() {
+            return Err(NumericError::Negative);
+        }
+
+        Ok(Self { inner })
+    }
+}
+
+impl core::str::FromStr for NumericSpec {
+    type Err = NumericSpecParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Valid formats:
+        // Numeric
+        // Numeric(scale)
```

Comment:
we don't allow setting precision?

Reply:
> No, i think it’s useless since we use fixed size representation for the Numeric type (and precision allow us control amount of space occupied by the type).

yes, atm. But would we want to optimize the storage it would be a breaking change. Let's discussion adding precision in a backwards compatible way in a separate ticket

Reply:
opened #4335